Source/Details/ASPageTable.h

```diff
@@ -0,0 +1,82 @@
+//
+//  ASPageTable.h
+//  Texture
+//
+//  Copyright (c) 2017-present, Pinterest, Inc.  All rights reserved.
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+
+#import <Foundation/Foundation.h>
+#import <UIKit/UIKit.h>
+#import <AsyncDisplayKit/ASBaseDefines.h>
+
+@class ASCollectionElement;
+
+NS_ASSUME_NONNULL_BEGIN
+
+AS_SUBCLASSING_RESTRICTED
+@interface ASPageCoordinates : NSObject
+
+@property (nonatomic, assign, readonly) NSUInteger x;
+@property (nonatomic, assign, readonly) NSUInteger y;
+
++ (instancetype)pageCoordinatesWithX:(NSUInteger)x Y:(NSUInteger)y;
+
+/**
+ * Returns the coordinates of the page that contains the specified point.
+ * Similar to CGRectContainsPoint, a point is considered inside the page if its coordinates lie inside the page or on the minimum X or minimum Y edge.
+ *
+ * @param point The point that the page at the returned coordinates should contain. Any negative coordinates of the point will be corrected to 0.0f
+ *
+ * @param pageSize
+ */
++ (instancetype)pageCoordinatesForPageThatContainsPoint:(CGPoint)point pageSize:(CGSize)pageSize;
+
+/**
+ * Returns the coordinates of all the pages that intersect the specified rect, restricted to the bounds of a content rect that has an origin of {0, 0} and a size of contentSize.
+ *
+ * @param rect The rect intersecting with the target pages.
+ *
+ * @param contentSize The combined size of all pages.
+ *
+ * @param pageSize The size of each page.
+ */
++ (NSArray<ASPageCoordinates *> *)pageCoordinatesForPagesInRect:(CGRect)rect withContentSize:(CGSize)contentSize pageSize:(CGSize)pageSize;
+
+- (CGRect)pageRectForPageSize:(CGSize)pageSize;
+
+@end
+
+typedef NSMapTable ASPageTable;
+
+/**
+ * A category for creating & using map tables meant for storing objects using ASPageCoordinates as keys.
+ */
+@interface NSMapTable (ASPageTableMethods)
+
+/**
+ * Creates a new page table with (NSMapTableStrongMemory | NSMapTableObjectPointerPersonality) for values.
+ */
++ (ASPageTable *)pageTableForStrongObjectPointers;
+
+/**
+ * Creates a new page table with (NSMapTableWeakMemory | NSMapTableObjectPointerPersonality) for values.
+ */
++ (ASPageTable *)pageTableForWeakObjectPointers;
+
+/**
+ * Builds a new page to layout attributes from the given layout attributes.
+ *
+ * @param layoutAttributes The layout attributes to build from
+ *
+ * @param pageSize The size of each page.
+ */
++ (ASPageTable<ASPageCoordinates *, NSMutableSet<UICollectionViewLayoutAttributes *> *> *)pageTableWithLayoutAttributes:(NSArray<UICollectionViewLayoutAttributes *> *)layoutAttributes pageSize:(CGSize)pageSize;
```

Comment:
It may not be worth it, but we could make the layoutAttributes argument `id<NSFastEnumeration>` and then pass the map table's object enumerator directly, rather than forming an array.